src/Transport/Receiving/RepeatedFailuresOverTimeCircuitBreaker.cs

```diff
@@ -23,54 +23,49 @@ public RepeatedFailuresOverTimeCircuitBreaker(string name, TimeSpan timeToWaitBe
 
         public void Success()
         {
-            var oldValue = Interlocked.Exchange(ref failureCount, 0);
+            var previousState = Interlocked.CompareExchange(ref circuitBreakerState, Disarmed, Armed);
 
-            if (oldValue == 0)
+            // If the circuit breaker was Armed or triggered before, disarm it
+            if (previousState == Armed || Interlocked.CompareExchange(ref circuitBreakerState, Disarmed, Triggered) == Triggered)
             {
-                return;
+                _ = timer.Change(Timeout.Infinite, Timeout.Infinite);
+                Logger.InfoFormat("The circuit breaker for {0} is now disarmed", name);
+                disarmedAction();
             }
-
-            timer.Change(Timeout.Infinite, Timeout.Infinite);
-            Logger.InfoFormat("The circuit breaker for {0} is now disarmed", name);
-            disarmedAction();
-            triggered = false;
         }
 
         public Task Failure(Exception exception, CancellationToken cancellationToken = default)
         {
-            lastException = exception;
-            var newValue = Interlocked.Increment(ref failureCount);
+            _ = Interlocked.Exchange(ref lastException, exception);
+
+            // Atomically set state to Armed if it was previously Disarmed
+            var previousState = Interlocked.CompareExchange(ref circuitBreakerState, Armed, Disarmed);
 
-            if (newValue == 1)
+            if (previousState == Disarmed)
```

Comment:
```suggestion
            if (Interlocked.CompareExchange(ref circuitBreakerState, Armed, Disarmed)== Disarmed)
```

Reply:
Plus, down here, `previousState` is used again on line 52, and it's good to have them both aligned for readability.

Reply:
Yeah, otherwise you would have to volatile read the `circuitBreakerState` again